Clipping planes

tschw opened the pull request:
~~Also includes my cleanup PR (#8464) - I will take it out, when it turns out necessary.~~
**Edit**: Took it out.

Adds a property to the render `.userClippingPlanes`, an array that can be filled with `THREE.Plane` objects. All fragments in one half-space defined by each plane are discarded, semantically pretty much like  [glClipPlane](https://www.opengl.org/sdk/docs/man2/xhtml/glClipPlane.xml) (present in fixed-function desktop OpenGL).

See #8462

tschw committed:
Implemented user-defined clipping planes.

tschw commented:
Example (not linked, because it's more of a test and not super pretty-looking) is `webgl_clipping_planes.html`.

erichlof commented:
@tschw This is awesome, thank you!

 @mrdoob We could really use this, as this is the best solution to clipping objects against an arbitrary user-defined plane.

WestLangley commented:
@tschw Sweet!

http://ci.threejs.org/api/pullrequests/8465/examples/webgl_clipping_planes.html

I think it would be particularly helpful at some point to show in the demo how to specify the clipping plane in world-space, camera-space, and object space.

tschw commented:
Talking about coordinate systems, shadows won't work correctly, I guess:

I do it all in view space (because minimum precision in the fragment shader is `mediump`), so needs a transform of the planes for every shadow camera.

WestLangley commented:
>I do it all in view space

I think in your example, the clipping planes are specified in world space. No?

tschw commented:
Yes, the specification is world space and then I do `tmp.copy(plane).applyMatrix(view)` and write the data into a flat array for the uniform array.

WestLangley commented:
>I think it would be particularly helpful at some point to show in the demo how to specify the clipping plane in world-space, camera-space, and object space.

Changed my mind. World-space only is adequate, IMO, which is what you have done.

EliasHasle commented:
Awesome! I was thinking about this just yesterday, and here it is! :-)

@WestLangley Shouldn't it be possible to clip a specific object only? Say, to clip a ship while leaving the ocean and skybox alone.

bhouston commented on examples/webgl_clipping_planes.html:
```diff
@@ -0,0 +1,94 @@
+				var N3 = Math.sqrt( 1 / 3 );
+				plane1 = new THREE.Plane( new THREE.Vector3( - N3, - N3, - N3 ), 120 );
+				plane2 = new THREE.Plane( new THREE.Vector3(   0 ,  N2 , - N2 ), 90 );
+				renderer.userClippingPlanes.push( plane1, plane2 );
```
I'd call it renderer.clippingPlanes?  I like the simple name.  Or customClippingPlanes if we need to be more specific.

tschw replied:
That's what I did. It's consistent and not so lengthy.

bhouston commented on src/materials/ShaderMaterial.js:
```diff
@@ -47,6 +47,7 @@ THREE.ShaderMaterial = function ( parameters ) {
 	this.fog = false; // set to use scene fog
 
 	this.lights = false; // set to use scene lights
+	this.userClippingPlanes = false; // set to use user-defined clipping planes
```
Hmmm.... are clipping planes material specific?  Or are they global renderer state?  I think this PR sort of confuses the two, or maybe that is on purpose?  I am okay with this but if we can make it one or the other, that would lead to a clearer design (unless this is needed.)

tschw replied:
It's built it in analogy to `ShaderMaterial.lights`: When you write a custom shader you state that you would like to get the uniform passed in - IOW it says whether your custom shader supports clipping.

I don't like this stuff so much either (am tired of ranting about it), been following what appears to be common practice, and would find it better to replace code like

```javascript
if ( material instanceof ThatMaterial ||
        material instanceof AnotherMaterial ||
        material instanceof YetAnotherMaterial ||
        material.supportsMyFeature ) {

      // feed uniform

}
```
with
```javascript
if ( material.supportsMyFeature ) {

    // feed uniform

}
```
and define `supportsMyFeature` on the prototypes of the materials - unless for the ones wrapping custom shaders where the property gets set on the instance.

MasterJames replied:
@tschw I agree with your comments about instanceof and I now prefer the function form of typeof as it gets away from none parentheses functions.
If ( typeof (1) === 'number' )
But sometimes you can't. LIke this is needed instead. for areays. 
Array.isArray (myArray) === true
So maybe you meant something different.

tschw replied:
> So maybe you meant something different.

I did. Meant what I wrote :-).

bhouston commented on src/renderers/shaders/ShaderChunk/clipping_planes_fragment.glsl:
```diff
@@ -0,0 +1,11 @@
+	for ( int i = 0; i < NUM_USER_CLIPPING_PLANES; ++ i ) {
+
+		vec4 plane = userClippingPlanes[ i ];
+		if ( dot( vViewPosition, plane.xyz ) > plane.w ) discard;
```
@tchw isn't there clipping plane support in the GL pipeline?  Argh, it isn't in OpenGL ES 2.

bhouston replied:
I'd suggest adding a math function for common.glsl so we can be a little higher level in this test.

```
if( planePointBehind( plane, vViewPosition ) discard;
```

I guess I was thinking we could create a math library in common.glsl (or split it out to geometry.glsl eventually)  once so we could use higher level functions in our glsl code.

tschw replied:
Is there clipping support in ES 3? I know there is in OpenGL 4... If so, we can just keep the infrastructure, and set an appropriate `gl_ClipDistance` at vertex level instead.

For a math library, it should probably be `signedPlaneDistance`, so the factorization has a factor > 1.

bhouston replied:
Feel free to add "signedPlaneDistance" to somewhere here:

https://github.com/mrdoob/three.js/blob/master/src/renderers/shaders/ShaderChunk/common.glsl#L58

bhouston replied:
OpenGL ES has an extension from Apple that adds clipping planes:

https://www.khronos.org/registry/gles/extensions/APPLE/APPLE_clip_distance.txt

I think glClipPlane was in openGL 1.x or something like that: https://www.khronos.org/opengles/sdk/1.1/docs/man/glClipPlane.xml

I think it got deprecated in favor of clip distance -- I am not familiar with these OpenGL 4.5 changes: http://web.eecs.umich.edu/~sugih/courses/eecs487/common/notes/APITables.xml

tschw replied:
`glClipPlane` is OpenGL < 3, or so - long ago.

`gl_ClipDistance` is the built-in vertex shader output that replaces it in modern desktop OpenGL versions.

tschw replied:
No built-in clipping support in ES 3 / WebGL2 :(.

tschw replied:
> Feel free to add "signedPlaneDistance" to somewhere here:

I prefer to feel even freer to leave it as is, because there is no more concise way to precisely specify the semantics than with a simple dot product. Citing the manual page of `glClipPlane`:

> If the dot product of the eye coordinates of a vertex with the stored plane equation components is positive or zero, the vertex is in with respect to that clipping plane. Otherwise, it is out.

So is this side the "negative half space", or how is it called? This way, I don't have to worry :-)

bhouston commented on src/renderers/shaders/ShaderChunk/clipping_planes_vertex.glsl:
```diff
@@ -0,0 +1,4 @@
+#if NUM_USER_CLIPPING_PLANES > 0 && ! defined( STANDARD ) && ! defined( PHONG )
+	vViewPosition = - mvPosition.xyz;
```
We maybe should just always pass world position into the fragment shader?  It is just a single and useful uniform.  That or have a way to reconstruct it.  Seems wasteful/complex that we maintain so many if tests to save a single very common uniform from being passed.

bhouston replied:
This does't have to be done as part of this PR -- just noticing it.

tschw replied:
> We maybe should just always pass world position into the fragment shader? It is just a single and useful uniform.

If you want it to run without warnings, the fragment shader must have a static data dependence on the uniform.

> That or have a way to reconstruct it. Seems wasteful/complex that we maintain so many if tests to save a single very common uniform from being passed.

I agree that the macro zoo is anything but pretty. In fact, I tidied more of it up more than I added back, yesterday. 

See #8464

bhouston replied:
So we could have fragment shaders scanned and any variable they reference then gets automatically created in the vertex shader if it didn't exist previously?  Sort of like fulfilling dependences?  That may be a way to deal with some of this macro zoo.

tschw replied:
I think that Three.js should undergo a similar transition as graphics APIs did when switching from fixed-function pipelines to programmable ones at some point in the (hopefully) not so distant future. Putting one clever patch on top of the other won't make it any less patchy and will only sustain our suffering. 

Thinking in terms of dependencies is certainly the right direction. Let there be data structures, generic data-driven machinery and well-defined interfaces for extension!
Scanning belongs into a build step, IMO. This way, it doesn't have to be super lightweight and can be done properly without bloating up the library core.

A dependency evaluator capable of piecing together programs from appropriately prepared GLSL parts on-demand (handling arbitrary graphs, including automatic synthesis of copy operations on forking connections, ordering constraints and error checking) can be implemented in just a few hundred bytes of JavaScript. A generic mechanism that determines the data sources for uniforms (object, material, traversal rule, ...) takes even less code. It all boils down to a few, well thought-out, yet in the end much simpler and easier to maintain key components that let us turn all the grown mess into data.

Retiring the macros and stuff  in `WebGLProgram(s)` plus resulting simplification of the renderer may easily free-up the space for it all - and the library will become exponentially more powerful. It should be possible for the user to hook into the default shading system, rewire or partially override it at any point between the blocks - instead of having to go through a lengthy and error-prone process of copy & paste and reverse-engineering of the glorious ShaderChunk framework.

Stuff for another PR. You know I love that topic.

bhouston commented:
Thank you @tschw.  My comments are not at all requirements, just random feedback.  This PR is amazing and we need to get it in.

tschw commented:
@WestLangley 
> > I think it would be particularly helpful at some point to show in the demo how to specify the clipping plane in world-space, camera-space, and object space.

> Changed my mind. World-space only is adequate, IMO, which is what you have done.

World space is the only thing that makes sense API-wise (unless we let the user specify the coordinate system, but that would be over-engineered IMO) and is fairly easy to reach from either of the spaces thanks to `Plane.applyMatrix`.

Use cases do exist in any of the spaces:

Model space: Cut an object for whatever reason - visualization, artistic stuff, or debugging.
World space: Portal culling (reminds me I should check the planes in bounding geometry tests as well, so it's performance-wise fit for that application).
View space: Removing the one-eye-clipped region before the convergence plane of a stereo viewport.

Shadows must be switchable. In some of these cases we'd want occluders to be cut, in others we wouldn't.

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -49,6 +49,10 @@ THREE.WebGLRenderer = function ( parameters ) {
 
+	// user-defined clipping
+
+	this.userClippingPlanes = [];
```
As far as I understand, this is mainly uniforms. Wouldn't it be better to set this array in the material instead?

tschw replied:
> As far as I understand, this is mainly uniforms. Wouldn't it be better to set this array in the material instead?

This array holds `THREE.Plane` objects in world space. 

Right now, clipping planes completely mask all rendering - not for a single object, not just for meshes: Points, sprites, meshes, whatever with a built-in shader. Clipping for shadow occlusion must be optional, as discussed above. 

The state is global and the resulting uniform value is shared between all materials - similar as we do for the lights (please note that the implementation of the update mechanics are still half-baked in the code here right now - I'm still hardening it).

There are several use cases where it would be cool to allow it on a per object basis. I don't think it really belongs onto the material, but it would be the "reasonably close approximation right now" to avoid issues like #8379 and keep this PR focussed (I guess that's why you are suggesting it?). 

Again, as with lighting, there is a conversion to view space in the renderer. This way, there are no problems when running on a platform with reduced fragment shader precision (floats scale nicely around the origin. When they are offset, the exponent gets defined by the translation and there are fewer bits to work with for fine-grained stuff. The nice thing about view space is, that we carry the origin around with us and everything closely visible is near the origin and gets decent precision). So there's a dependence on the camera and putting clipping planes on the material may end up a complication rather than a simplification.

Clipping planes effectively restrict the view frustum, so it's worthwhile to consider putting them on the camera object.

The dream API would have global clipping planes on the camera to mask the viewport  - ones that do not affect occluders when rendering a shadow map (just because you can't see the object doesn't mean you can't see its shadow). Then there could be additional ones on an object that semantically cut the object and therefore quite logically also its shadow.

One could then also unify frustum culling (should be a separate step) by always having six entries minimum - the clipping planes of the camera's view frustum.

tschw replied:
I'll have to investigate what is possible without major changes the renderer

tschw replied:
I left the planes array on the renderer for now (renamed to `.clippingPlanes`). I could fairly easily put the interface on the main camera, but I'm still undecided.

> There are several use cases where it would be cool to allow it on a per object basis. I don't think it really belongs onto the material, but it would be the "reasonably close approximation right now" to avoid issues like #8379 and keep this PR focussed (I guess that's why you are suggesting it?). 

I guess, I might have guessed wrong. Per-material planes would allow clipping a whole group of objects at once...

Now that the update paths are all working can shape it. In a while - I need to read over it again with a clear mind, maybe clean up a bit and think out how to best implement per-material clipping.

tschw commented:
Here is the revised changeset.

What's new?
- `.frustumCulled` objects are nor rendered when outside the clipping planes.
- `renderer.clipShadows` controls whether shadows are clipped or not.
- All updating works properly.

Fun test:
- Open `webgl_shadowmap_viewer`
- On the browser console:
```javascript
> plane = new THREE.Plane( new THREE.Vector3( 0, 0, -1 ), 1 )
> renderer.clippingPlanes.push( plane )
> renderer.clipShadows = true
> plane.constant = 0
```

WestLangley commented:
@EliasHasle 

>@WestLangley Shouldn't it be possible to clip a specific object only? Say, to clip a ship while leaving the ocean and skybox alone.

You should be able to do that with separate render passes, I would think.

@tschw 

Maybe we should have a `renderer.clippingPlanesEnabled` flag.

>Fun test:

: - )

tschw commented:
@EliasHasle 
> > Shouldn't it be possible to clip a specific object only? Say, to clip a ship while leaving the ocean and skybox alone.

@WestLangley 
> You should be able to do that with separate render passes, I would think.

I agree that it would be preferable to somehow be able to clip specific objects. Shadows would at least be tricky to get right with a multipass approach.

I think a `THREE.ClipGroup` that applies clipping to all contained objects in that scene graph node would probably be the best API for modeling purposes. These would be the ones that clip shadows (global clipping that does not affect shadows can be put onto the camera). 

It's currently still rather tricky to implement and better done after some internal issues are resolved (programs must be decoupled from the materials, see  #8379). Being able to set extra clipping planes on a material, as suggested by @mrdoob above, would be within reasonable reach, I guess.

@WestLangley 
> Maybe we should have a `renderer.clippingPlanesEnabled` flag.

What is your motivation of suggesting it? Efficiency or ease of use? Right now, it's technically `clippingPlanes.length` and its fine to change the identity of the array:
```javascript
renderer.clippingPlanes = [];
// ...
renderer.clippingPlanes = myPortal;
```
Efficiency-wise, I guess it would be nicer to wrap the code in `render` into `if` blocks. It's not a large overhead, but makes the source more readable.

https://github.com/tschw/three.js/blob/ClippingPlanes/src/renderers/WebGLRenderer.js#L1188-L1193
https://github.com/tschw/three.js/blob/ClippingPlanes/src/renderers/WebGLRenderer.js#L1204-L1209

However, an enable flag would require another test in the hot spot (that is frustum culling) with clipping enabled.

https://github.com/tschw/three.js/blob/ClippingPlanes/src/renderers/WebGLRenderer.js#L1346-L1361

So in case the suggestion is about efficiency, I'd opt for
```javascript
var clippingPlanesEnabled = this.clippingPlanes.length !== 0;
```
in `render` to wrap up the code.

mrdoob commented on src/renderers/shaders/ShaderLib/normal_vert.glsl:
```diff
@@ -3,6 +3,7 @@ varying vec3 vNormal;
 #include <common>
 #include <morphtarget_pars_vertex>
 #include <logdepthbuf_pars_vertex>
+#incldue <clipping_planes_pars_vertex>
```
typo: `#incldue`.

tschw replied:
Ouch! Thanks for spotting.

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1299,6 +1331,37 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	}
 
+	function objectCanBeVisible( object ) {
```
`function isObjectVisible( object )` maybe?

tschw replied:
I agree that the name is improvable. There are those `.visible` properties on objects and materials, so "visible" can mean a lot of things, and just because an object's bounding geometry intersects the frustum doesn't mean it's necessarily visible.

Something like `isNotCulled`, `isObjectNotCulled` or `objectIsNotCulled` might be even better. You like?

Note that one *could* make it public and call from `WebGLShadowMap` but I intentionally did not do this because:
- There are only clipping planes when `renderer.clipShadows === true`
- in this case, I interpret clipping as a modeling operation (in line with the typical use cases) - we're cutting objects, not visible space.

BTW (off-topic) https://2016.revision-party.net/  (Shader Live Coding Finale right now)

tschw replied:
I went with `isObjectViewable` to avoid confused with `.visible`.

WestLangley commented:
@tschw 

>Maybe we should have a renderer.clippingPlanesEnabled flag.
>>What is your motivation of suggesting it? 

It was a response to the use case specified by @EliasHasle above. Something like:
```javascript
ocean.layers.set( 0 );
ship.layers.set( 1 );
scene.add( ocean );
scene.add( ship );

...

camera.layers.set( 0 );
renderer.render( scene, camera );

camera.layers.set( 1 );
renderer.clippingEnabled = true;
renderer.render( scene, camera );
renderer.clippingEnabled = false;
```

bhouston commented:
BTW I'd advocate getting this in similar to what @tschw has implemented and then adding further complexity later.  I think he has done a good job on the implementation.  Yes, it isn't perfectly flexible for all cases but we can add that in later PRs.

tschw committed:
Renderer: Minor documentation update.

tschw committed:
Implements object-level clipping.

tschw committed:
Renderer: Minor cosmetic changes.

tschw committed:
Fixed lethal typo in normal shader.

tschw commented:
We have object-level clipping, now. API:
```javascript
renderer.clippingPlanes // global clipping planes, do not clip shadows
renderer.localClippingEnabled // enabled local clipping (default false for zero cost)
material.clippingPlanes // local clipping planes...
material.clipShadows // ...clip shadows when this one is set to true (default false)
shaderMaterial.clipping // requests the uniform to be fed to the custom shader
```

tschw commented:
@bhouston 
> BTW I'd advocate getting this in similar to what @tschw has implemented and then adding further complexity later. I think he has done a good job on the implementation. Yes, it isn't perfectly flexible for all cases but we can add that in later PRs.

I was about to suggest that, but it would have been half-baked for certain applications, so couldn't resist. I put some effort into having a clean internal encapsulation, to keep it adaptable to future changes.

Anyone know whether we have an accessible `cloneObjectArray` routine somewhere in the codebase to replace this code
https://github.com/tschw/three.js/blob/ClippingPlanes/src/materials/Material.js#L282-L295

??

tschw committed:
Let shadows respect .side of object's material.

tschw committed:
Fixed Plane.applyMatrix4.

- Scale denormalized but did not scale the position
- Projection is not supported (solved by documentation)
- Removed unnecessary copying to make up for the normalization
- Still arguably too heavy

tschw committed:
Added custom clipping example.

tschw commented:
New example:

http://ci.threejs.org/api/pullrequests/8465/examples/webgl_custom_clipping.html

tschw committed:
Clipping: Fixed transition to disabled state.

tschw committed:
Improved custom clipping example.

- GUI dependency of volume visualization on Enabled state
- Larger radius for cylindrical global clipping area

tschw commented:
Done pushing - feels complete.

erichlof commented:
@tschw That's beautiful!  Thank you for all of your efforts and expertise!  :-)

WestLangley commented:
Sweet!

One issue, though. The global clipping toggle does not appear to work correctly when local clipping is disabled.

tschw commented:
> One issue, though. The global clipping toggle does not appear to work correctly when local clipping is disabled.

That'd be fixed. I patched 0cb6b45 to e975d0e: 
```patch
diff --git a/src/renderers/WebGLRenderer.js b/src/renderers/WebGLRenderer.js
index 164dc0f..02b3674 100644
--- a/src/renderers/WebGLRenderer.js
+++ b/src/renderers/WebGLRenderer.js
@@ -2890,7 +2890,8 @@ THREE.WebGLRenderer = function ( parameters ) {
                                _this.clippingPlanes.length !== 0 ||
                                _this.localClippingEnabled ||
                                // enable state of previous frame - the clipping code has to
-                               // run another frame in order to reset the state
+                               // run another frame in order to reset the state:
+                               _numGlobalClippingPlanes !== 0 ||
                                _localClippingEnabled;
 
                _localClippingEnabled = _this.localClippingEnabled;
```

bhouston commented:
@tschw wrote:
> I was about to suggest that, but it would have been half-baked for certain applications, so couldn't resist. 

No worries, I love the extra contributions.  :)  Was just trying to make sure we were not making you run the gauntlet -- i think that sometimes talented people are made the run the gauntlet because they make for more engaging discussion or we know that they can code anything, so we ask them to.  But I think that this also discourages talented people from contributing.  Was just ensuring we avoid that scenario because I want clipping planes merged. :)

WestLangley commented on examples/webgl_custom_clipping.html:
```diff
@@ -0,0 +1,399 @@
+						multiply( tmpMatrix.makeRotationY( time * 0.2 ) );
+
+				object.matrix.copy( transform );
+				object.updateMatrixWorld( true );
```
You are overwriting the object matrix and setting `matrixAutoUpdate` to `false`. Do you mind changing this? We do not advise users doing that, and certain library methods do not work properly when that flag is `false`, anyway. Just set `object.rotation.x/y/z.`

tschw replied:
@WestLangley 

Ha! Yes, I thought that it might be a didactic mistake.

> Just set object.rotation.x/y/z.

OK. In this case I can use the object to compose the transform and then copy `.matrix` for the transform of the clipping volume.

However, I have to set the matrix for visualizing the planes (it'd be ill-advised to decompose it just for this purpose, I guess):

https://github.com/tschw/three.js/blob/ClippingPlanes/examples/webgl_custom_clipping.html#L256
https://github.com/tschw/three.js/blob/ClippingPlanes/examples/webgl_custom_clipping.html#L376-L377

What's the most advisable way to do so? I've actually been looking whether there is an `Object3D.setMatrix` routine... 

```javascript
object.matrix.makeIdentity();
object.applyMatrix( matrix );
```

Actually, since [this utility function](https://github.com/tschw/three.js/blob/ClippingPlanes/examples/webgl_custom_clipping.html#L92-L129) produces a world matrix, I might want to add another:

```javascript
function setObjectWorldMatrix( object, matrix ) {
    // cancel inherited world transform
    object.matrix.getInverse( object.parent.matrixWorld );
    // apply sought orientation
    object.applyMatrix( matrix );
}
```
The example does not need it, but this way it embeds more robust utility code for the user to copy over - and I don't have to touch `.matrixAutoUpdate`.

You like?

WestLangley replied:
We had many debates about what the user should and should-not be able to do years ago, so there is a lot of history regarding this issue. New users often don't know what they are doing. The policy has been to advise against setting the object matrix directly. That is the renderer's job.

That being said, I have no problem revisiting this issue if you feel you have a compelling reason to do so.

tschw replied:
> We had many debates about what the user should and should-not be able to do years ago, so there is a lot of history regarding this issue. New users often don't know what they are doing. The policy has been to advise against setting the object matrix directly. That is the renderer's job.

OK, I get that. I placed this example under "advanced". It's probably a bad choice for a 3D newbie to study. Then OTOH, the tricky bits are packaged for "faithful copy & paste". 

In some cases you need the matrices, and I believe it might be difficult to get away without in this one. There is a matrix interface, and of course I want to use it in a "generally advisable" fashion.
 
eadd4df gets rid of `.matrixAutoUpdate = false` and the transform of the main mesh is now calculated setting `.position` and `.rotation`. I want to apply the same translation / rotation to the clipping volume (an array of `Plane` objects) - and therefore I need a matrix, so I make sure the matrix is up to date and copy it ([see code](https://github.com/tschw/three.js/blob/ClippingPlanes/examples/webgl_custom_clipping.html#L362-L374)).

> That being said, I have no problem revisiting this issue if you feel you have a compelling reason to do so.

For visualizing the (world space) `Plane` objects as they are, I built a [utility function](https://github.com/tschw/three.js/blob/ClippingPlanes/examples/webgl_custom_clipping.html#L92-L129) that turns a plane into the (world) matrix for a `Mesh` with `PlaneGeometry`. It's fairly easy to do: An orthonormal space is constructed around `.normal` for the Z-Axis. [Another utility function](https://github.com/tschw/three.js/blob/ClippingPlanes/examples/webgl_custom_clipping.html#L131-L139) is provided to then align an object according to the matrix.

I can't really think of simpler means to do this stuff. In particular, I believe it would be ill-advised to encourage the user to use Euler angles for world space stuff, because they become unmanageable once there's a bit of scene hierarchy.

WestLangley replied:
@tschw 

>I placed this example under "advanced". It's probably a bad choice for a 3D newbie to study.

It doesn't have to be complicated. The features you have added are excellent, and your demo is visually appealing, but I agree, the code is too advanced for most users.

Users look to the examples for direction, and in many cases, they copy the code directly. Your example only needs one global clipping plane orthogonal to the world y-axis, and one local clipping plane orthogonal to the local y-axis -- perhaps each with a GUI slider. The clipping planes do not have to be visualized as planes, as they will be obvious. That would result in an easy-to-understand example for users to follow.

Please, note, however, this is just my opinion, and you are free to leave the example as-is if you prefer. : - )

tschw replied:
@WestLangley 
How about I add  `webgl_clipping_basic` and we rename this one to `webgl_clipping_advanced`?

Keep in mind that the planes are specified in world space and that `Plane` is a rather mathematical API. A very simple example leaves a lot of unanswered questions - especially for the non-experts. The example I have written shows how to work with spaces and planes, because ultimately that's what you gotta do in order to get predictable results. Point is, although admittedly hard-to-grok at noob level, there's a didactic value that I wouldn't want to throw away!

> The clipping planes do not have to be visualized as planes, as they will be obvious.

Not so much for a user who sees a black screen trying to set up multiple planes, happy to find a debugging tool in the example code :).

I also think you might be underestimating the average Three.js user. Just because the more experienced ones don't have to ask for support and just understand docs & code, doesn't mean they don't exist.

WestLangley replied:
>Keep in mind that the planes are specified in world space

Ah, yes, of course. Even for material-specific clip planes. For some reason I was thinking they were in local space.

> happy to find a debugging tool in the example code :).

That is a good, actually. Maybe `THREE.PlaneHelper` is in order?

>I also think you might be underestimating the average Three.js user. 

After answering ~ 1,000 three.js questions on stackoverflow, I do know that many users find the three.js examples too complicated to follow. Granted, that may be a biased sample, but I do favor simpler examples, in lieu of showcases. JMHO.

tschw committed:
No .matrixAutoUpdate = false in clipping example.

tschw commented:
@bhouston 
> Was just trying to make sure we were not making you run the gauntlet [...]  I think that this also discourages talented people from contributing

Wise words - kind & encouraging ones too. Thank you!

Having spent hours and hours debugging the state updating code paths in the Three.js renderer since I started contributing, I pretty much knew what I was getting myself into. Encouraging talented people to contribute is awesome - we could really use more hands and brains - ideally to reduce the number of lose ends in the codebase rather than to increase it, which is why I wanted it to be feature complete.

*Actually* this is a fairly simple thing - a one liner, if you will. Still, although the code has gotten quite a bit easier to grasp, encouraging the OP of #8462 to send a PR (like I usually do) would have been sadistic: Changing `WebGLRenderer`, and playing it safely within an (arguably) reasonable amount of time still takes one of the few people like you or me who are already familiar with the code. We are also automatically careful not to make too much of a mess (it's still pretty easy around there), because it will bite us soon enough - and it always takes more time than expected.

In this PR I actually took the questionable fame of finally adding my name to that file - as a matter of partial responsibility. Dozens of people have somehow hacked features into it, nobody seems to admit it, and when it all breaks apart we blame Ricardo for the merge :).

tschw committed:
Let ShadowMap check renderer.localClippingEnabled.

tschw commented on examples/webgl_custom_clipping.html:
```diff
@@ -0,0 +1,404 @@
+				// Geometry
+
+
+				var geometry = new THREE.TorusKnotGeometry( 0.3, 0.15, 60, 10 ); //new THREE.BoxGeometry( 0.6, 0.6, 0.6 );
```
Unused duplicate variable - leftover code.

tschw committed:
Added a simplified clipping example.

tschw commented:
As suggested [here](https://github.com/mrdoob/three.js/pull/8465#discussion_r57825398), there's now a simplified example:
http://ci.threejs.org/api/pullrequests/8465/examples/webgl_clipping.html

The other one has been renamed for reasons given [here](https://github.com/mrdoob/three.js/pull/8465#discussion_r57828599):
http://ci.threejs.org/api/pullrequests/8465/examples/webgl_clipping_advanced.html

WestLangley commented:
@tschw 

>As suggested here, there's another, simplified example now:

Ah, so fine!   : - )

tschw commented:
@WestLangley 
> Maybe `THREE.PlaneHelper` is in order?

I like the idea, but can we factorize the example in another PR? Also would need docs and some additional thought to be plug & play. In fact, one could imagine a `PlaneHelper` and a `ClippingVolumeHelper` - the latter visualizing a bunch of planes, each clipped against the respective others. Also one might want to have an offset parameter, so planes can be rendered with clipping enabled.

Some code may best go into some 'Utils' class or classes.

Construction of an orthonormal space from a single fixed axis is a useful utility in its own right, for instance. There's a newer (and faster) method than that of Hughes & Möller and I guess it might be less jumpy on the choice for the secondary axis. I used the simpler method for readability and because I knew that my planes would be blended. The jumpiness looks crappy when visible, e.g. moving wireframe planes...

WestLangley commented:
>I like the idea, but can we factorize the example in another PR?

That is fine with me.

Also, just to make sure you are aware, `Object3D.updateMatrixWorld()` doesn't actually do what it says. It should bubble up the scene graph, not down it. (`Scene.updateMatrixWorld()` is correct, however.) @gero3 invested a lot of effort in an approach to eliminate the need for the user to call that method by lazily updating the object- and world-matrix as needed. His PRs never were merged, however -- maybe due to some performance reasons.

gero3 commented:
The PRs @WestLangley is referring to are 
- #7019
- #6851
- #6448

The code behind it was working but I couldn't put much time in Three.js anymore. That is why it was abandoned. I'll see if I can revive the PR since we were close to a solution, both performance wise and API wise.

tschw commented:
> `Object3D.updateMatrixWorld()` doesn't actually do what it says. It should bubble up the scene graph, not down it [...] Scene.updateMatrixWorld() is correct

That sounds familiar - I think it bit me before. Code fixed in-place.

Anything else?

WestLangley commented:
@tschw Nothing else from me. Let's see what @mrdoob wants.

@gero3 Before you invest more of your time, it might be a good idea to check with @mrdoob again to see what his thinking is on the topic. I know you put a lot of effort into the issue.

tschw commented:
@gero3 
> > The code behind it was working but I couldn't put much time in Three.js anymore. That is why it was abandoned. I'll see if I can revive the PR since we were close to a solution, both performance wise and API wise. 

I'd be interested in a concluding summary of how the sought solution would look like. Would you mind opening a thread and posting a reference?

@WestLangley wrote:
> Before you invest more of your time, it might be a good idea to check with @mrdoob again to see what his thinking is on the topic. I know you put a lot of effort into the issue.

Also suits this purpose.

> Let's see what @mrdoob wants

Vanished from this thread - too noisy around here :(

/ping @mrdoob 

Latest example links again, because they're buried somewhere above:

http://ci.threejs.org/api/pullrequests/8465/examples/webgl_clipping.html
http://ci.threejs.org/api/pullrequests/8465/examples/webgl_clipping_advanced.html

tschw commented:
I'd like to take a shot at #8379, but it'd be easier to not have to manage concurrent changes in the renderer. Any chance to see this one merged soon?

/ping @mrdoob

WestLangley commented:
@tschw After merging :) ...

Global clipping does not appear to clip shadows -- but I expect it is not supposed to for the use cases envisioned?

Adding an option in the example for local clip planes to be defined in local space (rather than world space) would be nice. (I know the material sharing is an issue, but clipping in local space seems more natural to me. JMHO.)

	// clip in local space
	object.updateMatrixWorld();
	var clipPlane = object.material.clippingPlanes[ 0 ];
	clipPlane.copy( localPlane ).applyMatrix4( object.matrixWorld );

Your code works like a charm.

tschw commented:
@WestLangley 
> Global clipping does not appear to clip shadows -- but I expect it is not supposed to for the use cases envisioned?

Yes, it was intentional, as the typical use would be to clip the view.

> Adding an option in the example for local clip planes to be defined in local space (rather than world space) would be nice. (I know the material sharing is an issue, but clipping in local space seems more natural to me. JMHO.)

I want to get away from the material-based interface, once the material sharing issue has been resolved. I've been thinking about it for a while now, and believe I can pull it off in a nice way, that is also cleaning up the renderer. But that's clearly something for a different PR - after this one has been merged.

So,instead of improving the example, I'd like to improve the API. We could then have a `THREE.ClippingGroup` similar to `THREE.Group`, where the planes are then specified in that object's space. We could even go further and introduce a `THREE.ClippingPlane` which would be an `Object3D` that clips all the objects in the same `ClippingGroup`.

> `object.updateMatrixWorld();`

`scene.updateMatrixWorld()`, once again, no? :)

WestLangley commented:
@tschw 

>I want to get away from the material-based interface

Yes. That would be even better.

mrdoob commented:
Reading...

tschw commented:
`git rebase dev`

Is there actually any interest in merging this one?

bhouston commented:
I think this PR is very useful and valuable.  Thank you @tschw.  I've love to see it merged.

mrdoob commented:
> I want to get away from the material-based interface, once the material sharing issue has been resolved. I've been thinking about it for a while now, and believe I can pull it off in a nice way, that is also cleaning up the renderer. But that's clearly something for a different PR - after this one has been merged.

> So,instead of improving the example, I'd like to improve the API. We could then have a `THREE.ClippingGroup` similar to `THREE.Group`, where the planes are then specified in that object's space. We could even go further and introduce a `THREE.ClippingPlane` which would be an `Object3D` that clips all the objects in the same `ClippingGroup`.

That sounds great!

mrdoob commented:
Many thanks! And sorry for the delay 😇

mrdoob commented:
Btw, we should consider move all the relevant code to a `WebGLClipping` class. Otherwise we'll end up with a monter `WebGLRenderer` again 😀

tschw commented:
> Btw, we should consider move all the relevant code to a WebGLClipping class. Otherwise we'll end up with a monter WebGLRenderer again

I'd say `WebGLRenderer` still *is* a monster, although we try dieting it :).

I described what I'd like do in the next PR in #8552. The factorization would fit in there quite nicely, probably together with other render state stuff currently floating around in locals - we'll see how that one plays out.

arunkiitkgp commented:
@tschw  Hi I am using r77 version. I want to use tranformcontrols for clipping the object by translating or rotating the plane geometry against the object. But the problem is clippingPlanes is taking only THREE.Planes but not Three.Mesh (PlaneGeometry) for which transform controls are not applicable. Is there any solution so that i can translate Three.Plane using  transform controls.
It should look similar to the attachment which i have taken from google source. 
![3d tranform cont](https://cloud.githubusercontent.com/assets/8243617/16586687/bff5d9e8-42e4-11e6-8653-20aa3de31a64.PNG)

tschw commented:
@arunkkarepu 
Actually an SO question: A plane is a homogeneous row vector that is right-multiplied with a 4x4 transform, where x,y,z resemble the normal (from "normal direction of the plane", doesn't have to be normalized) and w is the "constant" (the origin offset along the normal given at similar scale).

For putting it together you may want to study the source of the `webgl_clipping_advanced` example.

patrickmatte commented:
Is it possible to make the material clippingPlanes property work with a ShaderMaterial?

tschw commented:
Yes, and there's even basic docs. See changes to `ShaderMaterial.html` in "Changed Files" of this issue and the shader code you'd have to add to your custom shader, of course. The feature has just been updated, so for the next generation of shader code see #9470.

patrickmatte commented:
Looked in "Changed files" and found a mention of the boolean clipping parameter but just adding that to my shader material didn't have any effect. I couldn't find the information about the shader code that I need to add to my shader. Do I need to add a uniform for the clipping planes?

tschw commented:
> Do I need to add a uniform for the clipping planes?

Yes, as for any engine functionality in Three.js, your custom shader needs to implement support for it.
